compiler/packages/babel-plugin-react-compiler/src/Inference/AnalyseFunctions.ts

```diff
@@ -110,84 +70,40 @@ function lower(func: HIRFunction): void {
   rewriteInstructionKindsBasedOnReassignment(func);
   inferReactiveScopeVariables(func);
-  inferMutableContextVariables(func);
   func.env.logger?.debugLogIRs?.({
     kind: 'hir',
     name: 'AnalyseFunction (inner)',
     value: func,
   });
 }
 
-function infer(
-  loweredFunc: LoweredFunction,
-  state: IdentifierState,
-  context: Array<Place>,
-): void {
-  const mutations = new Map<string, Effect>();
+function infer(loweredFunc: LoweredFunction): void {
+  const knownMutated = inferMutableContextVariables(loweredFunc.func);
   for (const operand of loweredFunc.func.context) {
-    if (
-      isMutatedOrReassigned(operand.identifier) &&
-      operand.identifier.name !== null
-    ) {
-      mutations.set(operand.identifier.name.value, operand.effect);
-    }
-  }
-
-  for (const dep of loweredFunc.dependencies) {
-    let name: IdentifierName | null = null;
-
-    if (state.properties.has(dep.identifier)) {
-      const receiver = state.properties.get(dep.identifier)!;
-      name = receiver.identifier.name;
-    } else {
-      name = dep.identifier.name;
-    }
-
-    if (isRefOrRefValue(dep.identifier)) {
+    const identifier = operand.identifier;
+    CompilerError.invariant(operand.effect === Effect.Unknown, {
+      reason:
+        '[AnalyseFunctions] Expected Function context effects to not have been set',
+      loc: operand.loc,
+    });
+    if (isRefOrRefValue(identifier)) {
       /*
        * TODO: this is a hack to ensure we treat functions which reference refs
        * as having a capture and therefore being considered mutable. this ensures
        * the function gets a mutable range which accounts for anywhere that it
        * could be called, and allows us to help ensure it isn't called during
        * render
        */
-      dep.effect = Effect.Capture;
-    } else if (name !== null) {
-      const effect = mutations.get(name.value);
-      if (effect !== undefined) {
-        dep.effect = effect === Effect.Unknown ? Effect.Capture : effect;
-      }
-    }
-  }
-
-  /*
-   * This could potentially add duplicate deps to mutatedDeps in the case of
-   * mutating a context ref in the child function and in this parent function.
-   * It might be useful to dedupe this.
-   *
-   * In practice this never really matters because the Component function has no
-   * context refs, so it will never have duplicate deps.
-   */
-  for (const place of context) {
-    CompilerError.invariant(place.identifier.name !== null, {
-      reason: 'context refs should always have a name',
-      description: null,
-      loc: place.loc,
-      suggestions: null,
-    });
-
-    const effect = mutations.get(place.identifier.name.value);
-    if (effect !== undefined) {
-      place.effect = effect === Effect.Unknown ? Effect.Capture : effect;
-      loweredFunc.dependencies.push(place);
+      operand.effect = Effect.Capture;
+    } else if (knownMutated.has(operand)) {
+      operand.effect = Effect.Mutate;
```

Comment:
This PR (the rewrite) preserves existing logic, but the `knownMutated` logic is removed altogether in https://github.com/facebook/react/pull/32097/